program/rust/src/tests/test_ema.rs

```diff
@@ -1,23 +1,133 @@
 extern crate test_generator;
 
 use {
+    super::test_utils::update_clock_slot,
     crate::{
-        accounts::PriceAccount,
+        accounts::{
+            PriceAccount,
+            PythAccount,
+        },
+        c_oracle_header::{
+            PC_STATUS_TRADING,
+            PC_VERSION,
+        },
+        deserialize::{
+            load_checked,
+            load_mut,
+        },
+        instruction::{
+            OracleCommand,
+            UpdPriceArgs,
+        },
         processor::{
             c_upd_aggregate,
             c_upd_twap,
+            process_instruction,
         },
+        tests::test_utils::AccountSetup,
     },
     bytemuck::Zeroable,
     csv::ReaderBuilder,
     serde::{
         Deserialize,
         Serialize,
     },
-    std::fs::File,
+    solana_program::pubkey::Pubkey,
+    std::{
+        fs::File,
+        mem::size_of,
+    },
     test_generator::test_resources,
 };
 
+#[test]
+fn test_ema_multiple_publishers_same_slot() -> Result<(), Box<dyn std::error::Error>> {
+    let mut instruction_data = [0u8; size_of::<UpdPriceArgs>()];
+    populate_instruction(&mut instruction_data, 10, 1, 1);
+
+    let program_id = Pubkey::new_unique();
+
+    let mut funding_setup = AccountSetup::new_funding();
+    let funding_account = funding_setup.as_account_info();
+
+    let mut price_setup = AccountSetup::new::<PriceAccount>(&program_id);
+    let mut price_account = price_setup.as_account_info();
+    price_account.is_signer = false;
+    PriceAccount::initialize(&price_account, PC_VERSION).unwrap();
+
+    {
+        let mut price_data = load_checked::<PriceAccount>(&price_account, PC_VERSION).unwrap();
+        price_data.num_ = 1;
+        price_data.comp_[0].pub_ = *funding_account.key;
+    }
+
+    let mut clock_setup = AccountSetup::new_clock();
+    let mut clock_account = clock_setup.as_account_info();
+    clock_account.is_signer = false;
+    clock_account.is_writable = false;
+
+    update_clock_slot(&mut clock_account, 1);
+
+    process_instruction(
+        &program_id,
+        &[
+            funding_account.clone(),
+            price_account.clone(),
+            clock_account.clone(),
+        ],
+        &instruction_data,
+    )?;
+
+    {
+        let price_data = load_checked::<PriceAccount>(&price_account, PC_VERSION).unwrap();
+        assert_eq!(price_data.prev_twap_.val_, 0);
+        assert_eq!(price_data.prev_twac_.val_, 0);
+        assert_eq!(price_data.twap_.val_, 10);
+        assert_eq!(price_data.twac_.val_, 1);
+    }
+
+    // add new test for multiple publishers and ensure that price_cumulative is updated correctly
```

Comment:
comment is wrong, we're not checking `price_cumulative`

Reply:
I don't get exactly what this test is testing wrt to this PR

Reply:
fixed the comment, can you give more contexts as to which specific part is unclear? this is to test that multiple publishers with price updates should lead to ema being updated